nucypher/blockchain/eth/clients.py

```diff
@@ -254,14 +284,84 @@ def block_number(self) -> BlockNumber:
+                        continue
+
+        else:
+            # If not asking for confirmations, just use web3 and assume the returned receipt is final
+            try:
+                receipt = self.w3.eth.waitForTransactionReceipt(transaction_hash=transaction_hash,
+                                                                timeout=timeout,
+                                                                poll_latency=self.TRANSACTION_POLLING_TIME)
+            except TimeExhausted:
+                raise  # TODO: #1504 - Handle transaction timeout
+
         return receipt
 
+    def block_until_enough_confirmations(self, transaction_hash: str, timeout: float, confirmations: int) -> dict:
+
+        receipt: TxReceipt = self.w3.eth.waitForTransactionReceipt(transaction_hash=transaction_hash,
+                                                                   timeout=timeout,
+                                                                   poll_latency=self.TRANSACTION_POLLING_TIME)
+
+        preliminary_block_hash = Web3.toHex(receipt['blockHash'])
+        tx_block_number = Web3.toInt(receipt['blockNumber'])
+        self.log.info(f"Transaction {Web3.toHex(transaction_hash)} is preliminarily included in "
+                      f"block {preliminary_block_hash}")
+
+        confirmations_timeout = self._calculate_confirmations_timeout(confirmations)
+        confirmations_so_far = 0
+        with Timeout(seconds=confirmations_timeout, exception=self.NotEnoughConfirmations) as timeout_context:
+            while confirmations_so_far < confirmations:
+                timeout_context.sleep(self.BLOCK_CONFIRMATIONS_POLLING_TIME)
+                self.check_transaction_is_on_chain(receipt=receipt)
+                confirmations_so_far = self.block_number - tx_block_number
+                self.log.info(f"We have {confirmations_so_far} confirmations. "
+                              f"Waiting for {confirmations - confirmations_so_far} more.")
+            return receipt
+
+    @staticmethod
+    def _calculate_confirmations_timeout(confirmations):
+        confirmations_timeout = 3 * AVERAGE_BLOCK_TIME_IN_SECONDS * confirmations
+        return confirmations_timeout
+
+    def check_transaction_is_on_chain(self, receipt: TxReceipt) -> bool:
+        transaction_hash = Web3.toHex(receipt['transactionHash'])
+        try:
+            new_receipt = self.w3.eth.getTransactionReceipt(transaction_hash=transaction_hash)
+        except TransactionNotFound:
+            reorg_detected = True
+        else:
+            reorg_detected = receipt['blockHash'] != new_receipt['blockHash']
+
+        if reorg_detected:
+            exception = self.ChainReorganizationDetected(receipt=receipt)
+            self.log.info(exception.message)
+            raise exception
+            # TODO: Consider adding an optional param in this exception to include extra info (e.g. new block)
+        return True
```

Comment:
This method never returns `False`, and only one thing we test - that there are no exceptions, so maybe remove return at all? just `-> None`?